### PR TITLE
chore: add validation to update/create APIs

### PR DESCRIPTION
**Pull request by kohlisid**
=== Test Update ISBSVC ===
update replica from 3 to 5
2 new pods are up
image

missing fields
image

wrong replica
image

### CODE REVIEW

**Comment by vigith** [comment]
how are we doing RBAC for this? it will only work with port-forward right?

File: server/apis/v1_1/handler.go [hunk 1]
```diff
@@ -347,31 +372,38 @@ func (h *handler) UpdateInterStepBufferService(c *gin.Context) {
...
+	requestBody, err := parseSpecFromReq(c, SpecTypeISB)
 	if err != nil {
 		errMsg := fmt.Sprintf("Failed to update the interstep buffer service: namespace %q isb-services %q: %v", c.Param("namespace"), c.Param("isb-services"), err.Error())
 		c.JSON(http.StatusOK, NewNumaflowAPIResponse(&errMsg, nil))
 		return
 	}
-
-	if requestBody.Redis != nil {
+	var updatedSpec = requestBody.(*dfv1.InterStepBufferService)
+	isValid := validateISBSpec(h, isbSVC, updatedSpec, ValidTypeUpdate)
```
Comment (jy4096) [inline server/apis/v1_1/handler.go:384]: I think for the isbsvc validation we can just keep this function and remove the other condition checks below... what do you think?

File: server/apis/v1_1/handler.go [hunk 2]
```diff
@@ -52,6 +53,11 @@ const (
 	SpecTypePatch    = "patch"
 )
 
+const (
+	ValidTypeCreate = "valid-create"
+	ValidTypeUpdate = "valid-update"
+)
```
Comment (yhl25) [inline server/apis/v1_1/handler.go:59]: make it a type `ValidType`?

File: server/apis/v1_1/handler.go [hunk 3]
```diff
@@ -168,7 +174,14 @@ func (h *handler) CreatePipeline(c *gin.Context) {
 	}
 	// Convert reqBody to pipeline spec
 	var pipelineSpec = reqBody.(*dfv1.Pipeline)
+	pipelineSpec.Namespace = ns
 
+	isValid := validatePipelineSpec(h, nil, pipelineSpec, ValidTypeCreate)
+	if isValid != nil {
+		errMsg := fmt.Sprintf("Failed to create pipeline, %v", isValid.Error())
+		c.JSON(http.StatusOK, NewNumaflowAPIResponse(&errMsg, nil))
+		return
```
Comment (yhl25) [inline server/apis/v1_1/handler.go:183]: I have introduced `respondWithError` method, please use that

File: server/apis/v1_1/handler.go [hunk 4]
```diff
@@ -168,7 +174,14 @@ func (h *handler) CreatePipeline(c *gin.Context) {
...
+	}
 	_, err = h.numaflowClient.Pipelines(ns).Create(context.Background(), pipelineSpec, metav1.CreateOptions{})
 	if err != nil {
 		errMsg := fmt.Sprintf("Failed to create pipeline %q, %v", pipelineSpec.Name, err.Error())
```
Comment (yhl25) [inline server/apis/v1_1/handler.go:187]:
```suggestion
	if err != nil {
		errMsg := fmt.Sprintf("Failed to create pipeline %q, %s", pipelineSpec.Name, err.Error())
```

File: server/apis/v1_1/handler.go [hunk 5]
```diff
@@ -228,20 +241,26 @@ func (h *handler) UpdatePipeline(c *gin.Context) {
 		NewNumaflowAPIResponse(&errMsg, nil)
 		return
 	}
-	pl, err := h.numaflowClient.Pipelines(ns).Get(context.Background(),
+	pipeSpec, err := h.numaflowClient.Pipelines(ns).Get(context.Background(),
 		pipeline, metav1.GetOptions{})
 	if err != nil {
-		errMsg := fmt.Sprintf("Failed to patch pipeline %q namespace %q, %v",
+		errMsg := fmt.Sprintf("Failed to update pipeline %q namespace %q, %v",
```
Comment (yhl25) [inline server/apis/v1_1/handler.go:247]:
```suggestion
		errMsg := fmt.Sprintf("Failed to update pipeline %q namespace %q, %s",
```

File: server/apis/v1_1/handler.go [hunk 6]
```diff
@@ -228,20 +241,26 @@ func (h *handler) UpdatePipeline(c *gin.Context) {
...
-	var pipelineSpec = reqBody.(*dfv1.Pipeline)
-	pl.Spec = pipelineSpec.Spec
-	fmt.Println("DEBUG", pl.ResourceVersion)
-	_, err = h.numaflowClient.Pipelines(ns).Update(context.Background(), pl, metav1.UpdateOptions{})
+	var updatedSpec = reqBody.(*dfv1.Pipeline)
+	pipeSpec.Spec = updatedSpec.Spec
+
+	isValid := validatePipelineSpec(h, pipeSpec, updatedSpec, ValidTypeUpdate)
+	if isValid != nil {
+		errMsg := fmt.Sprintf("Failed to update pipeline, %v", isValid.Error())
```
Comment (yhl25) [inline server/apis/v1_1/handler.go:259]:
```suggestion
	if isValid != nil {
		errMsg := fmt.Sprintf("Failed to update pipeline, %s", isValid.Error())
```

File: server/apis/v1_1/handler.go [hunk 7]
```diff
@@ -294,6 +313,12 @@ func (h *handler) CreateInterStepBufferService(c *gin.Context) {
 		return
 	}
 	var isbSpec = reqBody.(*dfv1.InterStepBufferService)
+	isValid := validateISBSpec(h, nil, isbSpec, ValidTypeCreate)
+	if isValid != nil {
+		errMsg := fmt.Sprintf("Failed to create interstepbuffer service spec, %v", isValid.Error())
+		c.JSON(http.StatusOK, NewNumaflowAPIResponse(&errMsg, nil))
+		return
```
Comment (yhl25) [inline server/apis/v1_1/handler.go:320]: `respondWithError`

File: webhook/validator/validator_test.go [hunk 8]
```diff
@@ -21,93 +21,17 @@ import (
...
-			},
-		},
-	}
-}
-
-func fakePipeline() *dfv1.Pipeline {
-	return &dfv1.Pipeline{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "test-pl",
-			Namespace: "test-ns",
```
Comment (yhl25) [inline webhook/validator/validator_test.go:81]: any specific reason for moving this code to utils?

**Comment by kohlisid** [comment]
Closing this PR
Would be done in #1098

**Pull request closed without merging**